### DISCUSSION TITLE
Color blending

### DISCUSSION THREAD

**Discussion by jac8888** [mikke89/RmlUi#207] [General]
Hello, I have an update on the color blending! ([A follow-up from here.](https://github.com/mikke89/RmlUi/pull/203))

I tried copying the method of blending from the animation code (ElementAnimation.cpp) into the gradient code (DecoratorGradient.cpp) and border code (GeometryBackgroundBorder.cpp) and tested it out. Personally I would recommend against using it in these places, or by default in the lerp.

The main reason is that OpenGL shaders blend in the simple way, which means that a rectangular box with colors only set for the corners will look different from a rounded box with colors set for the many points along the edge. A weaker reason that I'd suggest against it is that gradients in CSS blend in the simple way from what I can tell, so I think the other blending would be unexpected anyway.

To illustrate the first reason:

Rectangle
Colors are set on the four corners, and interpolation happens the shader, in the simple way. The gradient code makes no difference.
![square](https://user-images.githubusercontent.com/78047391/125989867-e7886351-6b06-4242-9868-d2941347b427.png)

Rounded corners, simple interpolation (the way it is now)
Some interpolation is done in the code to set the colors on many points along the edge of the circle. The interpolation is done in the same way it is done in the shader.
![simple_round](https://user-images.githubusercontent.com/78047391/125989891-e7e6a0c8-c8b5-4f1d-817a-0634b9d21d4b.png)

Rounded corners, algorithm from animation code
This is where the problem appears. The colors are set on the points along the edge of the circle, but the interpolation is not done in the same way as in the shader.
![incorrect_round](https://user-images.githubusercontent.com/78047391/125989978-6d7064f0-e193-4012-b82c-5c110ffe3baf.png)

Testing this, I noticed that this algorithm appears to be backwards. As you can see, it looks dark around the edges. To fix it, I've swapped the power and the square root. I will share this change later.

Here is a comparison I've done of the different algorithms.
![gradients](https://user-images.githubusercontent.com/78047391/125989914-0a24abdc-f064-4678-ad05-41836cb1de6d.png)

The first one is Mark's method from [here](https://stackoverflow.com/questions/22607043/color-gradient-algorithm).
The second one is simple blending.
The third one is the algorithm from the animation code.
The fourth one is the corrected version of the algorithm from the animation code.

Much better! Here is what the fixed version looks like in use.

![corrected](https://user-images.githubusercontent.com/78047391/125989992-9e0ad210-89fa-49a2-ab29-65209099078b.png)

However, it does look off in the middle, and the issue of square/rounded corners being different when used here does still remain.

![long](https://user-images.githubusercontent.com/78047391/125990021-6f2f7fb6-4bae-477d-8ebe-a93c4f44b1ce.png)

As for a rounding issue on the corrected version, there *is* a slight one that sometimes can be found when both endpoints are the same color. Visually, I wasn't able to notice it, and it seems much rarer than the one that happened before.
![rounding](https://user-images.githubusercontent.com/78047391/125990038-ce6e0cd2-12ca-4b8e-b3dc-c47cf9399f9c.png)
![selection](https://user-images.githubusercontent.com/78047391/125990062-50f5e39e-ea22-4672-906f-ff6b860d94c6.png)

In conclusion, I think that the interpolation in the gradient and border code should continue to use the simple method of blending. I don't think the rounding error is much of an issue in the animation code, and I think it is reasonable to use the fixed version of that algorithm there.

**Comment by mikke89** [comment]
Wow, thanks a lot for this well-researched post. Impressive and well written!

All of this reminded me about a post I read a while back about [how CSS color interpolation is wrong by design](https://erikmcclure.com/blog/everyone-does-srgb-wrong-because/).

It's a bit challenging for RmlUi as a library since we intentionally don't control the color interpolation on the backend renderer, this is fully controlled by the user. You're right that with our sample shell the proper approach would be to do the naïve blending approach as we currently do. On the other hand, some engines may do a "more correct" color blending in the backend, and the naïve interpolation will then look wrong if it doesn't match the method used.

Here is one example from my own custom engine written in OpenGL:

![color](https://user-images.githubusercontent.com/5490330/126040467-042a3711-3c00-483a-8a39-c4cd2c40040e.png)

5 and 6 are rendered using the current master branch of RmlUi, while 1-4 are from your stack overflow reference. As can be seen here, the current RmlUi interpolation creates some issues when combined with a renderer which does color interpolation in linear space, there it would probably look better with the corrected animation approach. However our naïve sample shell backend looks correct as you found.

The main issue is that there is no "correct" way to do color interpolation, it's all subjective after all. I do think Mark's method looks the best out of those four, but I doubt any game engines would want to do all that math, sounds a bit too expensive.

I think you might be right, let's just do the naïve thing for now because everyone else does it. We might consider making a way to control this interpolation in the future?

And thanks a lot for finding the reversed conversion in the animation color interpolation, that indeed looks a whole lot better!

**Reply by jac8888** [reply]
Thanks for your response! You make a great point. Now I want to change my back end and keep the nice interpolation, haha. (To tell the truth, I had forgotten it was possible and an option... 😅) I tried it with the animation approach and it looks really nice. This is the outcome:

![better_square](https://user-images.githubusercontent.com/78047391/126045906-fac207ed-f2ad-4ace-ad85-d999987af34a.png)
![better_round](https://user-images.githubusercontent.com/78047391/126045911-931637cb-7822-4532-8be8-aa1e9cb138a3.png)
![better_rectangle](https://user-images.githubusercontent.com/78047391/126045922-90ed5f30-279a-4a9a-9c45-824a7c5f6182.png)
![better_rectangle_round](https://user-images.githubusercontent.com/78047391/126045925-c96e3427-84a4-4289-9957-c16472294e4f.png)

Considering it that way, I like the idea of somehow being able to use it.

I wonder if there's any interest in pushing the gradient completely into the shader (supplying the end colors and angle), taking less control rather than more? I'm not sure how feasible the move would be, and it would put more of a burden on people making their back ends, but it would allow for more flexibility, and more easily allow gradients at arbitrary angles.

**Reply by mikke89** [reply]
That looks good indeed!

I'd prefer not to add special render commands to the API like that. We already have custom decorators which users can use for more specialized rendering. It would be nice with rotation and other color gradient modes, but I'd prefer to do that with vertices and the current API.

What we could do is add a run-time (or compile-time) option to select between the two blending modes. Primarily the naïve method in nonlinear space or the inverted sRGB linear blending. I think a global option should suffice in 99% of use cases. What do you think about that? If so, I think we should use the actual sRGB gamma conversion rather than the simplified one from the animation interpolation.

**Reply by jac8888** [reply]
Haha, yeah, that's definitely understandable. In that case, I agree that a global option to choose between those two would work very well!
